### ISSUE TITLE
Maybe a request?

### ISSUE THREAD

**Issue by bradjc** [thejpster/tockloader-proto-rs#2]
Hi @thejpster this is very cool. I'd like to get us using this as soon as possible.

Can I make a hopefully simple request? It would be great if there were send and receive interfaces that used slices in addition to the individual byte interfaces. The c bootloader we are using now uses UART on a byte-by-byte basis, but the UART driver in Tock (https://github.com/helena-project/tock/blob/master/kernel/src/hil/uart.rs#L57) uses a slice interface.

My attempt at a bootloader built on Tock is here https://github.com/helena-project/tock/tree/bootloader. Right now it has a "new" board (hail-bootloader) and just a start on a bootloader capsule. We have support for writing flash and UART in Tock, and with this module we have code for the interface. Getting a first pass Tock based bootloader should be a matter of stitching things together.

**Comment by thejpster** [comment]
So, I think all you need to do for sending bytes is allocate a buffer of a size of your choosing, iterate on the Encoder until it returns None (or your buffer is full) and then pass the buffer to your UART. Likewise on the receive, you receive into a buffer, then loop through that buffer passing one byte at a time into the decoder until it returns `Ok(Some(...))`. I could add functions which do this for you, but the trick is working out how to tell the caller how much of the buffer has been read/written to if/when an error occurs or a command/response has been found.

**Comment by bradjc** [comment]
I think on the parsing side when a command comes in from the UART it's fine to assume there will only be one possible packet in the buffer and that it will start at the beginning of the slice, and that it is either a valid packet or not. The return should just be the parsed command or an error, the length doesn't really matter. On the encoding side, the return I think could just be the length of the response message or an error (which I guess would only occur if the buffer wasn't large enough?).

I do think that working with buffers is a useful interface for the bootloader protocol library particularly for chips that have DMA support.

**Comment by thejpster** [comment]
Humm. I can implement a function on the Encoder structs which takes a buffer to write to, but I can't with the Decoder structs. It basically gives this error:

https://play.rust-lang.org/?gist=93f5b110198a63265e4b2cb43a5585f6&version=stable

Thoughts?

**Comment by niklasad1** [comment]
I think it is enough to put the struct data in an container with interior mutability such as Cell or RefCell
And take `&self` as argument instead of `&mut self` because it looks like you try to borrow the struct mutably twice.

**Comment by alevy** [comment]
The problem there is that the example in the playground reuses the internal buffer for the `Command`, so the mutable reference to `self` in `receive` is actually held until the return value (`Result<Option<Command>, ()>`) falls out of scope.

This might work? https://play.rust-lang.org/?gist=ceabda51d6f99a376c41c7b57ec1a8db&version=stable

**Comment by alevy** [comment]
To clarify, it uses a separate buffer that's allocated _inside_ the `Command` type. In this case it's allocated on the stack, and it's small (between the buffer and the variant, `Command` would fit in two registers), but if this pattern needs to be used for much larger buffers, some more gymnastics may be required.

**Comment by thejpster** [comment]
So the problem with that is the actual code holds a 512 byte flash page, and I was trying not to use too much stack (or use Box, because then I need an allocator).

**Comment by thejpster** [comment]
Random thought. If you can guarantee the buffer contains a whole command, I can return a reference to the buffer passed in, rather than the decoder's internal buffer. Can you make that guarantee, or does your UART just return whatever it has available?

**Comment by thejpster** [comment]
Second thought - I can make this work if you're happy to receive the decoded `Command` via a callback. This convinces borrowck that the `&Command` is dropped before the next run of the loop.

**Comment by thejpster** [comment]
See https://github.com/thejpster/tockloader-proto-rs/commit/f9d7d4e3853db8c4bf3d9348f2f99bca2decc3e3.

**Comment by thejpster** [comment]
Let me know if this works and I'll `cargo publish` 0.2

**Comment by bradjc** [comment]
The two issues I'm running into:

1. How do I get the buffer back after passing it to `decoder.read()`?
2. I need a buffer interface for response encoded to send back the reply.

**Comment by thejpster** [comment]
I'm not sure I follow. The `read` function takes a slice, so it shouldn't claim any kind of ownership of your buffer.

The `CommandEncoder` and `ResponseEncoder` both have a `write` method (it's in the `Encoder` trait) which iterates over the output bytes and puts them in the given buffer. Unfortunately I didn't make the `Encoder` trait public, so you couldn't see it. See version 0.2.1 for the fix.

**Comment by thejpster** [comment]
@bradjc, is there anything else I can do to help here?

**Comment by bradjc** [comment]
Hi @thejpster. We finally got to building a bootloader based on Tock: https://github.com/tock/tock-bootloader/pull/9 and this library was very helpful for parsing commands send to the bootloader.

Unfortunately I found that handling responses didn't match well because of the need to pass in multiple pointers to buffers for responses like `ReadAttribute`, so the responses I just implemented directly in the bootloader.

If you have time/interest, take a look. Maybe there is a way the use of this library could be improved.
